### ISSUE TITLE
Gopher support

### ISSUE THREAD

**Issue by dertuxmalwieder** [OtterBrowser/otter-browser#1050]
I'd like to request support for the Gopher protocol, maybe as a module. I am willing to help to contribute it myself if I figure out how Otter determines a page's protocol... :)

**Comment by Emdek** [comment]
@dertuxmalwieder, why not, I'm not familiar with gopher but it seems that it will require translation of replies into HTML plus some other stuff, so probably it would be best to have it as some sort of module that would be used by web backends for producing HTML etc.
We could use it as example for other protocols in future.

**Comment by dertuxmalwieder** [comment]
Actually, it's not even "some other stuff". :-)

**Comment by Emdek** [comment]
@dertuxmalwieder, so basically just detecting reply that needs to be transformed and translating it?

**Comment by dertuxmalwieder** [comment]
Yep. You'll send a very basic "request" to the Gopher server (usually running on port 70), i.e. the URL, and it will reply with a "Gopher page" that just needs to be translated into HTML. Should be achievable with simple search&replace.

[Wikipedia example](https://en.wikipedia.org/wiki/Gopher_%28protocol%29#Protocol). The first line is the request, the rest is the Gopher page, each line being "an item".

**Comment by Emdek** [comment]
@dertuxmalwieder, I see (in fact I've visited that article while determining fate of this ticket).
If you are still willing to give it a try then I think that the easiest way to create initial implementation (specific to QtWebKit) would be to hook it here:
https://github.com/OtterBrowser/otter-browser/blob/master/src/modules/backends/web/qtwebkit/QtWebKitNetworkManager.cpp#L693
Take a look at implementation of `LocalListingNetworkReply` and `QtWebKitFtpListingNetworkReply` (these two are a bit similar since both deal with file listing, you can also use some sort of template file if it makes sense) or at least read documentation of `QNetworkReply`.
We can later try to create full blown generic implementation that could be supported also in QtWebEngine backend.

**Comment by dertuxmalwieder** [comment]
Looks like I could - more or less - copy the FTP implementation and rip out most of its commands. :-)

I'm still willing to give it a try, of course. I haven't had much spare time yet (and Otter keeps evolving)... but it's still on my list.

**Comment by Emdek** [comment]
@dertuxmalwieder, yeah, we can discuss details on IRC when you will have some time.
One of the topics I would like to care take of in next weeks would be enabling protocol handlers configuration (currently destined to become Programs section in Advanced, but that might change during redesign needed by transition to **about:preferences**) so it could be useful to take such handler modules into account.

**Comment by dertuxmalwieder** [comment]
I'll be on the IRC in approx. eight hours. I'll ping you. :)

**Comment by Emdek** [comment]
@dertuxmalwieder, OK, but eight hours from now might be a bit late to catch me on IRC...

**Comment by dertuxmalwieder** [comment]
Well, I randomly see you talking when I'm online. Don't worry, I'll get you some day.

**Comment by Emdek** [comment]
@dertuxmalwieder, OK.

**Comment by smaragdus** [comment]
Any development about adding [Gopher](https://en.wikipedia.org/wiki/Gopher_%28protocol%29) support to [Otter](https://otter-browser.org/)? I have just checked [Dooble](https://textbrowser.github.io/dooble/) (a browser which I do not like) and as of version **2020.07.07** Gopher is still supported. I wonder whether some part of [Dooble](https://github.com/textbrowser/dooble) code might be used to implement Gopher support in [Otter](https://github.com/OtterBrowser/otter-browser).

**Comment by dertuxmalwieder** [comment]
Very different code bases make that relatively hard.

**Comment by Emdek** [comment]
Patches are welcome. ;-)
